src/linker/Linker/UnconditionalSuppressMessageAttributeState.cs

```diff
@@ -0,0 +1,109 @@
+﻿using System.Collections.Generic;
+using Mono.Cecil;
+
+namespace Mono.Linker
+{
+	public class UnconditionalSuppressMessageAttributeState
+	{
+		private readonly LinkContext _context;
+		private readonly Dictionary<IMetadataTokenProvider, Dictionary<int, SuppressMessageInfo>> _localSuppressionsByMdToken;
+
+		private bool HasLocalSuppressions {
+			get {
+				return _localSuppressionsByMdToken.Count != 0;
+			}
+		}
+
+		public UnconditionalSuppressMessageAttributeState (LinkContext context)
+		{
+			_context = context;
+			_localSuppressionsByMdToken = new Dictionary<IMetadataTokenProvider, Dictionary<int, SuppressMessageInfo>> ();
+		}
+
+		public void AddLocalSuppression (CustomAttribute ca, IMetadataTokenProvider mdTokenProvider)
+		{
+			SuppressMessageInfo info;
+			if (!TryDecodeSuppressMessageAttributeData (ca, out info)) {
+				return;
+			}
+
+			if (!_localSuppressionsByMdToken.TryGetValue (mdTokenProvider, out var suppressions)) {
+				suppressions = new Dictionary<int, SuppressMessageInfo> ();
+				_localSuppressionsByMdToken.Add (mdTokenProvider, suppressions);
+			}
+
+			if (suppressions.ContainsKey (info.Id))
+				_context.LogMessage (MessageContainer.CreateInfoMessage (
+					$"Type or member {mdTokenProvider} has more than one unconditional suppression. Note that only the last one is used."));
+
+			suppressions[info.Id] = info;
+		}
+
+		public bool IsSuppressed (int id, MessageOrigin warningOrigin, out SuppressMessageInfo info)
+		{
+			info = default;
+
+			if (HasLocalSuppressions && warningOrigin.MdTokenProvider != null) {
+				IMetadataTokenProvider mdTokenProvider = warningOrigin.MdTokenProvider;
+				while (mdTokenProvider != null) {
+					if (IsLocallySuppressed (id, mdTokenProvider, out info))
+						return true;
+
+					mdTokenProvider = (mdTokenProvider as IMemberDefinition).DeclaringType;
+				}
+			}
+
+			return false;
+		}
+
+		private static bool TryDecodeSuppressMessageAttributeData (CustomAttribute attribute, out SuppressMessageInfo info)
+		{
+			info = default;
+
+			// We need at least the Category and Id to decode the warning to suppress.
+			// The only UnconditionalSuppressMessageAttribute constructor requires those two parameters.
+			if (attribute.ConstructorArguments.Count < 2) {
+				return false;
+			}
+
+			// Ignore the category parameter because it does not identify the warning
+			// and category information can be obtained from warnings themselves.
+			// We only support warnings with code pattern IL####.
+			if (!(attribute.ConstructorArguments[1].Value is string warningId) ||
+				warningId.Length < 6 ||
+				!warningId.StartsWith ("IL") ||
+				!int.TryParse (warningId.Substring (2, 4), out info.Id)) {
+				return false;
+			}
+
+			if (warningId.Length > 6 && warningId[6] != ':')
+				return false;
+
+			if (attribute.HasProperties) {
```

Comment:
Where are the tests for this logic?

Reply:
For now I've only added an attribute with these properties set to `null` in the tests. Most interesting tests can be implemented after we integrate global suppressions (which will need to use the signature parser for resolving the `Target` and `MessageId`).